File: Python/partition-string.py
# Time:  O(n)
# Space: O(t)

# simulation, trie
class Solution(object):
    def partitionString(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        class Trie(object):
            def __init__(self):
                self.__nodes = []
                self.__new_node()
                self.__curr = 0
            
            def __new_node(self):
                self.__nodes.append([-1]*26)
                return len(self.__nodes)-1

            def add(self, c):
                x = ord(c)-ord('a')
                if self.__nodes[self.__curr][x] == -1:
                    self.__nodes[self.__curr][x] = self.__new_node()
                    self.__curr = 0
                    return
                self.__curr = self.__nodes[self.__curr][x]

            def curr(self):
                return self.__curr

        result = []
        trie = Trie()
        curr = []
        for x in s:
            curr.append(x)
            trie.add(x)
            if trie.curr():
                continue
            result.append("".join(curr))
            curr = []
        return result

File: C++/partition-string.cpp
// Time:  O(n)
// Space: O(t)

// simulation, trie
class Solution {
public:
    vector<string> partitionString(string s) {
        vector<string> result;
        Trie trie;
        string curr;
        for (const auto& c : s) {
            curr.push_back(c);
            trie.add(c);
            if (trie.curr()) {
                continue;
            }
            result.emplace_back(move(curr));
        }
        return result;
    }

private:
    class Trie {
    public:
        Trie() {
             new_node();
         }

        void add(int c) {
            const auto& x = c - 'a';
            if (nodes_[curr_][x] == -1) {
                nodes_[curr_][x] = new_node();
                curr_ = 0;
                return;
            }
            curr_ = nodes_[curr_][x];
        }

        int curr() const {
            return curr_;
       }

    private:
        int new_node() {
            nodes_.emplace_back(26, -1);
            return size(nodes_) - 1;
        }

        vector<vector<int>> nodes_;
        int curr_ = 0;
    };
};
